warehouse/packaging/metadata_verification.py

```diff
@@ -41,6 +56,109 @@ def _verify_url_pypi(url: str, project_name: str, project_normalized_name: str)
+
+    Raises `GetContentError` on any errors while trying to
+    access the URL.
+    """
+    try:
+        http = urllib3.HTTPSConnectionPool(
+            host=resolved_ip,
+            port=443,
+            headers={
+                "Host": url.host,
+                "User-Agent": "warehouse/1.0 (PyPI URL verifier, https://pypi.org)",
+            },
+            server_hostname=url.host,
+            assert_hostname=url.host,
+            retries=False,
+            timeout=5,
+        )
+        r = http.request("GET", url.path if url.path else "/", preload_content=False)
+    except Exception:
+        raise GetContentError()
+
+    content = r.read(max_length_bytes)
+
+    # When using `preload_content=False`, the HTTP connection must be manually closed
+    r.close()
+
+    return content
+
+
+def _verify_url_meta_tag(
+    url: str, project_name: str, project_normalized_name: str
+) -> bool:
+    user_uri = rfc3986.api.uri_reference(url).normalize()
+
+    # Require the presence of "https" and the host name. Don't allow
+    # the presence of ports
+    if user_uri.port is not None:
+        return False
+    validator = (
+        rfc3986_validators.Validator()
+        .require_presence_of("scheme", "host")
+        .allow_schemes("https")
+    )
+
+    try:
+        validator.validate(user_uri)
+    except rfc3986ValidationError:
+        return False
+
+    # Only allow registered names for the host (e.g.: google.com, api.github.com),
+    # don't allow IP addresses.
+    if IPv4_MATCHER.fullmatch(user_uri.host) or IPv6_MATCHER.fullmatch(user_uri.host):
+        return False
+
+    # The domain name should not resolve to a private or shared IP address
+    try:
+        address_tuples = socket.getaddrinfo(user_uri.host, user_uri.port)
+        if len(address_tuples) == 0:
+            return False
+        for family, _, _, _, sockaddr in address_tuples:
+            ip_address: ipaddress.IPv4Address | ipaddress.IPv6Address | None = None
+            if family == socket.AF_INET:
+                ip_address = ipaddress.IPv4Address(sockaddr[0])
+            elif family == socket.AF_INET6:
+                ip_address = ipaddress.IPv6Address(sockaddr[0])
+            if ip_address is None or not ip_address.is_global:
+                return False
+    except (socket.gaierror, ipaddress.AddressValueError):
+        return False
+
+    # Use the first IP address returned by `getaddrinfo`
+    _, _, _, _, resolved_sockaddr = address_tuples[0]
+    resolved_ip = resolved_sockaddr[0]
+
+    try:
+        content = _get_url_content(
+            resolved_ip=resolved_ip,
+            url=user_uri,
+            max_length_bytes=_MAX_RESPONSE_LENGTH_BYTES,
+        )
+    except GetContentError:
+        return False
+
+    try:
+        html_root = html.document_fromstring(content)
+    except (StopIteration, etree.ParserError):
+        return False
+
+    meta_tag = html_root.xpath("//head/meta[@namespace='pypi.org' and @rel='me']")
```

Comment:
```suggestion
    meta_tag = html_root.xpath("//head/meta[@name='pypi-project' and @rel='me']")
```
`namespace` doesn't seem to be a valid attribute per https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/meta.

Additionally, making this something that looks kind of like a URL potentially opens this up to unnecessary user confusion (e.g. putting `https://pypi.org` or `pypi.org/` or something else).

Finally, making the name just `pypi.org` might prevent us from using meta tags for other types of validation. It might not be necessary, but we could easily future-proof this a bit by making the name specific to project verification

Reply:
Ah yeah, good catch -- technically this could be `data-namespace` to be compliant, but I don't think there's any good reason to not use `name` 🙂 

(FWICT `rel="me"` is also not documented on `<meta>`, but seems to be widely used with `<link>`. So maybe we should use `<link>` for this? OTOH there's no actual "link" per se.)